### PR TITLE
feature: add Canny edge detection operator

### LINKED ISSUE #40: Add Canny Edge Detection Operator
**Difficulty:** Medium
**Area:** backend, frontend

## Description

Canny edge detection is one of the most widely used edge detection algorithms in computer vision. It produces clean, thin edges by applying a multi-stage process: noise reduction via Gaussian blur, gradient computation, non-maximum suppression, and hysteresis thresholding. Adding this operator to ImageLab will give users a fundamental tool for feature extraction and image analysis workflows.

The operator should accept two threshold parameters that control the hysteresis step. Pixels with gradient magnitudes above `threshold2` are considered strong edges, those below `threshold1` are discarded, and those in between are kept only if connected to strong edges. The operator must also handle color images gracefully by converting them to grayscale before applying the Canny algorithm, since `cv2.Canny()` requires single-channel input.

## Acceptance Criteria

- [ ] A Canny edge detection operator is implemented in the backend following the existing operator pattern
- [ ] The operator handles color images gracefully (converts to grayscale internally) and outputs a pipeline-compatible image
- [ ] Two threshold parameters (threshold1, threshold2) are exposed with sensible defaults
- [ ] The operator is registered and available in the processing pipeline
- [ ] A Blockly block exists with two numeric fields for the thresholds, appearing in the filtering category
- [ ] End-to-end: adding the block to a pipeline and running it produces visible edge output in the preview pane

## Resources

- [OpenCV Canny Edge Detection docs](https://docs.opencv.org/4.x/da/d22/tutorial_py_canny.html)
- [cv2.Canny API reference](https://docs.opencv.org/4.x/dd/d1a/group__imgproc__feature.html#ga04723e007ed888ddf11d9ba04e2232de)
- Existing filtering operators in `imagelab-backend/app/operators/filtering/` for implementation patterns

### PR DESCRIPTION
**Pull request by xxivani**
## Description

Implements the Canny edge detection operator for ImageLab. Adds a new backend operator that accepts two threshold parameters (threshold1, threshold2) for hysteresis thresholding and handles color images by converting to grayscale internally before applying cv2.Canny(). Also adds the corresponding Blockly block to the frontend filtering category.

Fixes #40 

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?
Manually tested end to end by building a pipeline with Read Image → Canny Edge Detection and running it in the preview pane. The processed output correctly shows white edges on a black background for a color input image.
- [ ] Existing tests pass
- [ ] New tests added
- [x] Manual testing

## Screenshots 
Screenshot 2026-02-27 at 4 10 35 PM

## Checklist
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [ ] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [ ] Tests pass locally

### CODE REVIEW

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

The Canny edge detection feature is well-structured and follows existing codebase patterns cleanly — the color-space handling logic, BGR output normalization, and frontend block definition are all solid. However, there are two backend bugs that could cause crashes or silently incorrect output: `cv2.Canny()` only accepts `uint8` images (a float input from a prior pipeline step will throw a `cv2.error` at runtime), and there is no guard ensuring `threshold1 <= threshold2` (swapped values invert the hysteresis logic without raising an error). Additionally, the PR removes s390x platform wheels from `uv.lock` without explanation — this is an unrelated and potentially breaking infrastructure change that should not be bundled here. No automated tests were added for the new operator.

File: imagelab-frontend/src/blocks/definitions/filtering.blocks.ts [hunk 1]
```diff
@@ -112,4 +112,19 @@ export const filteringBlocks = [
     style: "filtering_style",
     tooltip: "Applies image sharpening to enhance edges and details",
   },
+  {
+    type: "filtering_cannyedge",
+    message0: "Canny Edge Detection %1 Threshold 1 %2 %3 Threshold 2 %4",
+    args0: [
+      { type: "input_dummy" },
+      { type: "field_number", name: "threshold1", value: 100, min: 0, max: 255 },
+      { type: "input_dummy" },
```
Comment (ivantha) [inline imagelab-frontend/src/blocks/definitions/filtering.blocks.ts]: **[LOW]** Frontend threshold max of 255 is overly restrictive for Canny

Both `threshold1` and `threshold2` are capped at `max: 255` in the Blockly field definitions. However, `cv2.Canny()` thresholds represent gradient *magnitudes*, not pixel intensities. For a standard uint8 image processed by a Sobel-based Canny, gradient magnitudes can realistically reach ~360 at a sharp 0→255 edge. This means a threshold of 255 would already be near the maximum useful value and users exploring higher thresholds (common when tuning detection sensitivity on high-contrast images) will be silently prevented from doing so by the UI.

This mismatch is also inconsistent: the backend imposes no upper limit, but the frontend does.

**Suggestion:**
```
Increase the maximum to a more permissive value, such as 500 or 1000:
```typescript
{ type: "field_number", name: "threshold1", value: 100, min: 0, max: 500 },
{ type: "input_dummy" },
{ type: "field_number", name: "threshold2", value: 200, min: 0, max: 500 },
```
This aligns with common Canny usage guides and removes an artificial constraint the backend does not enforce.
```

File: imagelab-backend/pyproject.toml [hunk 2]
```diff
@@ -40,3 +40,8 @@ packages = ["app"]
 [build-system]
 requires = ["hatchling"]
 build-backend = "hatchling.build"
+
+[dependency-groups]
+dev = [
+    "ruff>=0.15.4",
+]
```
Comment (ivantha) [inline imagelab-backend/pyproject.toml]: **[LOW]** Duplicate dev dependency declaration — PEP 735 group alongside existing optional-dependencies

The PR adds a `[dependency-groups]` section (PEP 735, supported by `uv` ≥ 0.4) to declare `ruff` as a dev dependency. However, the lock file already contains a `provides-extras = ["dev"]` entry, indicating `ruff` was already declared under `[project.optional-dependencies]` dev. This results in two parallel mechanisms declaring the same dependency:

1. `[project.optional-dependencies]` dev (traditional, pip-compatible)
2. `[dependency-groups]` dev (PEP 735, uv-native)

This is confusing for contributors and could cause `ruff` to be installed twice or create unexpected resolution conflicts depending on the tool used.

**Suggestion:**
```
Decide on one canonical approach for dev dependencies. If the project has fully committed to `uv`, migrate entirely to `[dependency-groups]` and remove the `[project.optional-dependencies]` dev extra. If pip compatibility is needed, keep only `[project.optional-dependencies]`. Do not use both simultaneously.
```

File: imagelab-backend/app/operators/filtering/canny_edge.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import cv2
+import numpy as np
+
+from app.operators.base import BaseOperator
+
+
+class CannyEdge(BaseOperator):
+    def compute(self, image: np.ndarray) -> np.ndarray:
```
Comment (ivantha) [inline imagelab-backend/app/operators/filtering/canny_edge.py]: **[LOW]** Missing class and method docstrings

The `CannyEdge` class and its `compute` method have no docstrings. Other operators in the codebase likely have docstrings (especially if auto-documentation is generated). More importantly, the class docstring would be the natural place to document the behaviour contract: that color images are converted to grayscale internally, that the output is always 3-channel BGR, and the semantics of `threshold1` and `threshold2`.

**Suggestion:**
```
```python
class CannyEdge(BaseOperator):
    """Applies Canny edge detection to an image.

    Color (BGR/BGRA) images are converted to grayscale before processing.
    The output is always a 3-channel BGR image with white edges on a black
    background, suitable for downstream pipeline operators.

    Params:
        threshold1 (float): Lower hysteresis threshold (default: 100).
        threshold2 (float): Upper hysteresis threshold (default: 200).
            Must be >= threshold1.
    """

    def compute(self, image: np.ndarray) -> np.ndarray:
        ...
```
```

File: imagelab-backend/app/operators/filtering/canny_edge.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        threshold1 = float(self.params.get("threshold1", 100))
+        threshold2 = float(self.params.get("threshold2", 200))
+
+        if len(image.shape) == 3:
+            if image.shape[2] == 4:
+                image = cv2.cvtColor(image, cv2.COLOR_BGRA2GRAY)
+            else:
+                image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
```
Comment (ivantha) [inline imagelab-backend/app/operators/filtering/canny_edge.py]: **[LOW]** Unhandled edge case: images with channel count other than 1, 3, or 4

The color conversion block handles `shape[2] == 4` (BGRA) and implicitly treats everything else as BGR. However, it does not guard against unusual channel counts (e.g., a 2-channel image from a prior pipeline operator). If `image.shape[2] == 2`, `cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)` will raise a `cv2.error` because the conversion code for a 2-channel input to grayscale doesn't exist. While this is an edge case, it produces an opaque OpenCV error rather than a helpful diagnostic.

**Suggestion:**
```
Add an explicit guard for unexpected channel counts:
```python
if len(image.shape) == 3:
    channels = image.shape[2]
    if channels == 4:
        image = cv2.cvtColor(image, cv2.COLOR_BGRA2GRAY)
    elif channels == 3:
        image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    else:
        raise ValueError(
            f"Unsupported number of channels: {channels}. Expected 3 (BGR) or 4 (BGRA)."
        )
```
```

File: imagelab-backend/app/operators/registry.py [hunk 5]
```diff
@@ -85,6 +86,7 @@
     "filtering_erosion": Erosion,
     "filtering_dilation": Dilation,
     "filtering_morphological": Morphological,
+    "filtering_cannyedge": CannyEdge,
```
Comment (ivantha) [inline imagelab-backend/app/operators/registry.py]: **[INFO]** Registry key naming: 'filtering_cannyedge' vs 'cannyedge' casing — minor consistency question

The registry key `"filtering_cannyedge"` follows the same pattern as `"filtering_erosion"` and `"filtering_dilation"`, which is consistent. However, `cannyedge` (no separator between the two words) is slightly harder to read than, say, `canny_edge`. Existing keys like `"filtering_morphological"` are single words so no precedent exists for multi-word subtypes. This is a very minor point, but if the project ever adds more operators with multi-word names, establishing a convention now avoids inconsistency later.

**Suggestion:**
```
If adding any convention, consider documenting that compound names use camelCase or underscores (e.g. `filtering_canny_edge`). The current `filtering_cannyedge` is acceptable if the project intends to avoid extra underscores, but it's worth a brief note in the PR.
```

**Comment by xxivani** [comment]
Hi @ivantha!
The branch has been rebased into main and all findings have been addressed.

File: imagelab-backend/app/operators/filtering/canny_edge.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    """
+
+    def compute(self, image: np.ndarray) -> np.ndarray:
+        threshold1 = max(0.0, float(self.params.get("threshold1", 100)))
+        threshold2 = max(0.0, float(self.params.get("threshold2", 200)))
+
+        # Enforce hysteresis ordering
+        if threshold1 > threshold2:
+            threshold1, threshold2 = threshold2, threshold1
+
```
Comment (ivantha) [inline imagelab-backend/app/operators/filtering/canny_edge.py]: **[LOW]** Silent threshold swap may surprise users; not reflected in UI tooltip

When `threshold1 > threshold2`, the values are silently swapped:

```python
if threshold1 > threshold2:
    threshold1, threshold2 = threshold2, threshold1
```

This is a helpful UX safeguard, but it is **not mentioned in the Blockly block tooltip**. A user who accidentally sets Threshold 1 higher than Threshold 2 gets valid (but unexpected) output with no feedback. The block tooltip only explains the hysteresis semantics, not this auto-correction.

The discrepancy between the Python docstring (which does document the swap) and the UI tooltip is a minor inconsistency.

**Suggestion:**
```
Append to the block tooltip:

```ts
tooltip:
  "Applies Canny edge detection ... Color images are automatically converted to grayscale. "
  + "If Threshold 1 is set higher than Threshold 2, the values are automatically swapped.",
```

Alternatively, consider emitting a warning log on the backend when a swap occurs, so pipeline debugging is easier.
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

The implementation of the Canny edge detection operator is clean and well-structured — good docstring, hysteresis threshold swap guard, multi-channel dispatch, and dtype safety are all positive. However, there is one clear high-severity bug: a duplicate `filtering_cannyedge` entry in `categories.ts` (one newly added, one pre-existing stub) that will render the block twice in the UI. Two medium-severity issues also warrant attention before merge: the single-channel `(H, W, 1)` image shape is not handled and will raise rather than gracefully proceeding, and float images normalised to `[0.0, 1.0]` will be silently clipped to near-black before edge detection. The absence of automated tests is a medium concern given the number of edge-case input paths. All changes are purely additive with no breaking surface.

**Suggested Version Bump:** `minor`

### ⚠️ Breaking Changes

- **[INFO]** No breaking changes — purely additive feature: **Previous behavior**: The `filtering_cannyedge` operator did not exist in the registry or block definitions.

**New behavior**: A new `CannyEdge` class is registered under `"filtering_cannyedge"` and a corresponding Blockly block is added to the filtering category.

**Who is affected**: No existing consumers are affected. No existing exports were removed or renamed. No API endpoints, schemas, or configuration formats were changed.

**Migration path**: None required. This is a backward-compatible addition.
  > **Suggestion:** No action needed. Suggested semver bump is `minor` (new, non-breaking feature added).


### PR Description Issues

- **[MEDIUM]** Multiple unchecked test checklist items with no explanation: The PR description leaves three test-related checklist items unchecked with no accompanying justification:

```
- [ ] Existing tests pass
- [ ] New tests added      (acknowledged)
- [ ] Tests pass locally
```

"Existing tests pass" and "Tests pass locally" are not just aspirational — they are gates for merging. Leaving them unchecked signals that CI was not run or tests were not executed, which reduces reviewer confidence. The PR description should either confirm the test suite was run (and note any pre-existing failures if the CI is broken), or explain why it could not be run.
  > **Suggestion:** Update the checklist section with explicit status:

```markdown
## How Has This Been Tested?
Manually tested end-to-end: Read Image → Canny Edge Detection pipeline produces
white edges on a black background for a color input image (see screenshot).

Automated tests for the new operator are not included in this PR (tracked in #XX).
The existing test suite passes locally with no regressions.

- [x] Existing tests pass
- [ ] New tests added — follow-up issue #XX
- [x] Tests pass locally
```
- **[LOW]** Description lacks motivation and use-case context: The description explains *what* was implemented (Canny operator with two thresholds, grayscale conversion, Blockly block) but does not explain *why* it was needed or what user workflows it enables beyond the bare mechanics. For a first-time reviewer unfamiliar with issue #40, there is no answer to: "What was the user-facing gap this fills?"

Additionally, the description does not mention the pre-existing `filtering_cannyedge` stub entry that appears to have already existed in `categories.ts` (visible as unchanged context in the diff), which should have been noted as a cleanup task.
  > **Suggestion:** Add a **Motivation** paragraph:

```markdown
## Motivation
Canny is one of the most widely used edge detection algorithms and a common
precursor to contour analysis, feature extraction, and document scanning
pipelines. ImageLab currently offers Sobel/Laplacian but lacks a
hysteresis-based edge detector. This closes the gap raised in #40 and enables
users to build pipelines such as:
  Read Image → Canny Edge Detection → Contour Detection
```

Also note the cleanup of the pre-existing stub entry.

**Comment by ivantha** [comment]
CI is currently red (failing check: Backend Lint). Could you take a look and push a fix? Happy to merge once CI is green.

**Comment by xxivani** [comment]
Hi @ivantha, I've pushed the fix now!

**Comment by ivantha** [comment]
Thanks @xxivani! Closing this in favour of #316 which implements the same Canny Edge Detection operator as a focused PR with CI passing. Appreciate the early work on this feature!
